tests/src/JIT/HardwareIntrinsics/X86/Shared/AlternatingBinOpTest_DataTable.cs

```diff
@@ -38,8 +38,8 @@ public AlternatingBinaryOpTest__DataTable(TOp1[] inArray1, TOp2[] inArray2, TRes
 
             this.simdSize = unchecked((byte)(simdSize));
 
-            Unsafe.CopyBlockUnaligned(ref Unsafe.AsRef<byte>(inArray1Ptr), ref Unsafe.As<TOp1, byte>(ref inArray1[0]), this.simdSize);
-            Unsafe.CopyBlockUnaligned(ref Unsafe.AsRef<byte>(inArray2Ptr), ref Unsafe.As<TOp2, byte>(ref inArray2[0]), this.simdSize);
+            Unsafe.CopyBlockUnaligned(ref Unsafe.AsRef<byte>(inArray1Ptr), ref Unsafe.As<TOp1, byte>(ref inArray1[0]), inArray1.Length * sizeof(TOp1));
```

Comment:
It's not obvious to me why the old code was bad and the new code is correct, the array length is `simdSize * 2` so it seems that copying `simdSize` bytes should have worked correctly. Besides, I doubt that `sizeof(TOp1)` will compile.

Reply:
The problem was the size of arguments (`inArray1`, `inArray2`, `outArray`, etc) could be different.

> the array length is simdSize * 2 

`simdSize * 2` is the length of buffers (the members) rather than input arrays (the arguments). 

> Besides, I doubt that sizeof(TOp1) will compile.

I tried, it works well.

Reply:
The code in this file is plain bogus...

Reply:
😂

Reply:
> simdSize * 2 is the length of buffers (the members) rather than input arrays (the arguments).

Right. And that is odd since the buffers end up being smaller than the input arrays. I assume that `simdSize` is the size of a vector.

Reply:
> that is odd since the buffers end up being smaller than the input arrays

No, the problem was `simdSize` may be larger than the input arrays when copying the input to buffers.

> simdSize is the size of a vector.

Actually, it is the largest alignment boundary.

Reply:
Oops, `sizeof(TOp1)` does not work...

Reply:
> Actually, it is the largest alignment boundary.

Isn't that the same thing as the vector size? SSE2 vector size is 16 bytes and the required alignment is 16 bytes. `CopyBlockUnaligned` copies bytes the old code copied 16 bytes. How could the input arrays be smaller than 16 bytes?

Reply:
For any given test, `simdSize` should be the size of the largest `Vector` used (so 16 or 32, currently).

We then allocate `2 * simdSize` so that we can guarantee the buffer is aligned (since you need at most `alignment - 1` additional bytes to guarantee alignment).

In the case where one argument is `Vector128` and one argument is `Vector256`, simdSize will be larger than the input array for the `Vector128` args, and will result in an AV.

Reply:
Except you don't usually get an AV. The code just keeps copying whatever random bytes there are past the end of the array. So the failure only shows up in GC stress when we break in just after the copy and find there is a byref and it happens to be pointing into free space.

Reply:
> In the case where one argument is Vector128 and one argument is Vector256, simdSize will be larger than the input array for the Vector128 args, and will result in an AV.

I see. Well, this code is a mess. 

* The `DataTable` name is misleading, this whole class isn't some sort of data table, it's just some kind of wrapper for some aligned memory buffers (and it's not clear why instead of simply having a `AlignedArray<T>` class we have instead this set of classes where the logic is duplicated).
* The `simdSize` parameter should have been called alignment, that's its main purpose here.
* There are no asserts that would suggest that the input arrays are supposed to contain only one vector and thus be smaller than `simdSize`.
* In the absence of such asserts the length of the new arrays should have been calculated based on the parameter arrays, not based on `simdSize`.
* The out array parameter is not used
* The way alignment is done is sort of unusual. Pretty much every alignment function I've seen does `(x + (alignment - 1)) & ~(alignment - 1)`.
* It probably would have been simpler to just allocate memory using `Marshal.AllocCoTaskMem` instead of dragging in GC handles as well

The actual test code isn't any better. Tons of repeated code, unsafe copies and whatnot. I take it that people were in a hurry to get things done and this is the result.